experimental/lib/Support/BooleanLogic/Lexer.cpp

```diff
@@ -0,0 +1,110 @@
+//===- Lexer.cpp - Lexer for boolean expressions ----------------*- C++ -*-===//
+//
+// Dynamatic is under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements the LexicalAnalyzer class for tokenizing a string
+// representing a boolean logic expression. The constructor processes the input
+// string, identifying tokens and storing them in a list. The GetToken method
+// retrieves the next token, while peek allows looking ahead in the token
+// stream.
+//
+//===----------------------------------------------------------------------===//
+
+#include "experimental/Support/BooleanLogic/Lexer.h"
+#include "llvm/Support/raw_ostream.h"
+#include <string>
+#include <utility>
+#include <vector>
+
+using namespace llvm;
+using namespace dynamatic::experimental::boolean;
+
+// Function for reporting syntax errors encountered during lexical analysis.
+void dynamatic::experimental::boolean::syntaxError() {
+  llvm::errs() << "Syntax error in expression!\n";
+}
+
+LexicalAnalyzer::LexicalAnalyzer(std::string exp) {
+  expression = std::move(exp);
+}
+
+// The function tokenize processes a string expression, skipping whitespace and
+// identifying variable tokens from letter-digit sequences. It creates tokens
+// for operators and parentheses, storing all tokens in tokenList. An end token
+// is added to mark the completion of tokenization.
+void LexicalAnalyzer::tokenize() {
+  size_t position = 0;
+  while (position < expression.length()) {
+    char current = expression[position];
+    if (isspace(current)) {
+      position++;
+      continue;
+    }
+    if (isalpha(current)) { // If the current character is a letter, it
+                            // indicates a variable.
+      size_t idIndex = position + 1;
+      while (idIndex < expression.length() && isdigit(expression[idIndex]))
+        idIndex++;
+      std::string id = expression.substr(position, idIndex - position);
+      Token token(id, TokenType::VariableToken);
+      tokenList.push_back(token);
+      position = idIndex;
+
+    } else {
+      Token token;
+      std::string str(1, current);
+      token.lexeme = str;
+      switch (current) {
+      case '.':
+      case '&':
+        token.tokenType = TokenType::AndToken;
+        break;
+      case '+':
+      case '|':
+        token.tokenType = TokenType::OrToken;
+        break;
+      case '~':
+      case '!':
+        token.tokenType = TokenType::NotToken;
+        break;
+      case '(':
+        token.tokenType = TokenType::LparenToken;
+        break;
+      case ')':
+        token.tokenType = TokenType::RparenToken;
+        break;
+      case '0':
+      case '1':
+        token.tokenType = TokenType::VariableToken;
+        break;
+      default:
+        syntaxError();

```

Comment:
You should stop parsing and return a failure in that case (`LogicalResult`), it's not enough to write to `stderr`.

Reply:
Would you recommend returning a pair<LogicalResult, BoolExpression*> in the parse function? Or a variant maybe?

Reply:
If the function is supposed to return a non-null `BoolExpression*` then you can return `nullptr` as a "failure value"; this is a consise and typical way for functions returning pointers to signal failure.

Reply:
If the function does not semantically have a return value beyond a success/failure state then `LogicalResult` is more appropriate.